src/families/cardano/js-signOperation.ts

```diff
@@ -0,0 +1,175 @@
+  accountKey: Bip32PublicKey,
+  witnesses: Array<Witness>
+) => {
+  witnesses.forEach((witness) => {
+    const [, , , chainType, index] = witness.path;
+    const publicKey = accountKey
+      .derive(chainType)
+      .derive(index)
+      .toPublicKey()
+      .toBytes();
+    const vKeyWitness: TyphonTypes.VKeyWitness = {
+      signature: Buffer.from(witness.witnessSignatureHex, "hex"),
+      publicKey: Buffer.from(publicKey),
+    };
+    unsignedTransaction.addWitness(vKeyWitness);
+  });
+
+  return unsignedTransaction.buildTransaction();
+};
+
+/**
+ * Sign Transaction with Ledger hardware
+ */
+const signOperation = ({
+  account,
+  deviceId,
+  transaction,
+}: {
+  account: Account;
+  deviceId: string;
+  transaction: Transaction;
+}): Observable<SignOperationEvent> =>
+  new Observable((o) => {
+    async function main() {
+      const transport = await open(deviceId);
+      try {
+        o.next({ type: "device-signature-requested" });
+
+        if (!transaction.fees) {
+          throw new FeeNotLoaded();
+        }
+
+        const unsignedTransaction = await buildTransaction(
+          account,
+          transaction
+        );
+
+        const accountPubKey = getExtendedPublicKeyFromHex(
+          account.xpub as string
+        );
+
+        const rawInputs = unsignedTransaction.getInputs();
+        const ledgerAppInputs = rawInputs.map((i) =>
+          prepareLedgerInput(i, account.index)
+        );
+
+        const rawOutptus = unsignedTransaction.getOutputs();
+        const ledgerAppOutputs = rawOutptus.map((o) =>
+          prepareLedgerOutput(o, account.index)
+        );
+
+        const trxOptions: SignTransactionRequest = {
+          signingMode: TransactionSigningMode.ORDINARY_TRANSACTION,
+          tx: {
+            network:
+              CARDANO_NETWORK_ID === Networks.Mainnet.networkId
+                ? Networks.Mainnet
+                : Networks.Testnet,
+            inputs: ledgerAppInputs,
+            outputs: ledgerAppOutputs,
+            certificates: [],
+            withdrawals: [],
+            fee: unsignedTransaction.getFee().toString(),
+            ttl: unsignedTransaction.getTTL()?.toString(),
+            validityIntervalStart: null,
+            auxiliaryData: null,
+          },
+          additionalWitnessPaths: [],
+        };
+
+        // Sign by device
+        const appAda = new Ada(transport);
+        const r = await appAda.signTransaction(trxOptions);
+        const signed = signTx(unsignedTransaction, accountPubKey, r.witnesses);
+        o.next({ type: "device-signature-granted" });
+
+        const operation = buildOptimisticOperation(
+          account,
+          transaction.amount,
+          unsignedTransaction
+        );
+
+        o.next({
+          type: "signed",
+          signedOperation: {
+            operation,
+            signature: signed.payload,
+            signatureRaw: signed,
+            expirationDate: null,
+          },
```

Comment:
Please type this object `as SignedOperation`, to make sure it has the correct shape. I'm only asking because in other integrations some sneaky bugs were caused by this object not being aligned with what's used in `broadcast`. Unfortunately we lack a way to enforce this implicitly...